### PR TITLE
bpo-36071 Add support for Windows ARM32 in ctypes/libffi

### PR DESCRIPTION
**Pull request by paulmon**
This will need to rebased on the pending libffi PR before it is merged
All test_ctypes tests pass on both retail and debug builds except for test_SEH (retail only)
I am trying to see if I can get help with the unwind semantics in arm assembly using MSVC tools
Also https://github.com/python/cpython-source-deps/pull/9 will need to be merged before this will build. 

https://bugs.python.org/issue36071

### CODE REVIEW

**Comment by paulmon** [comment]
Submitted ARM32 changes to libffi: https://github.com/libffi/libffi/pull/477

**Comment by zooba** [comment]
Since I assume this PR is going to require new libffi builds, keep using the branch name in get_externals until it's ready. Then remind me to tag it and remind you to change to use the tag before merging :)

**Comment by zooba** [comment]
It's probably also a good idea to rebase this onto master and merge all the commits (do a merge, then `git reset --soft upstream/master`, re-commit, and force push).

**Comment by paulmon** [comment]
@zooba I cleaned up this PR.  

I also cleaned up the PR for the arm32 libffi-7.dll changes:
https://github.com/python/cpython-source-deps/pull/9

I have build this locally and tested test_ctypes on a Raspberry Pi 3 with success.
I kicked off all of the standard library tests and will check on that later.

File: Include/internal/pycore_atomic.h [hunk 1]
```diff
@@ -407,13 +407,13 @@ typedef struct _Py_atomic_int {
...
-    _InterlockedExchange_acq((volatile long*)&((ATOMIC_VAL)->_value), (int)NEW_VAL); \
+    _InterlockedExchange_acq((volatile long*)(ATOMIC_VAL), (int)NEW_VAL); \
     break; \
   case _Py_memory_order_release: \
-    _InterlockedExchange_rel((volatile long*)&((ATOMIC_VAL)->_value), (int)NEW_VAL); \
+    _InterlockedExchange_rel((volatile long*)(ATOMIC_VAL), (int)NEW_VAL); \
     break; \
   default: \
-    _InterlockedExchange((volatile long*)&((ATOMIC_VAL)->_value), (int)NEW_VAL); \
+    _InterlockedExchange((volatile long*)(ATOMIC_VAL), (int)NEW_VAL); \
```
Comment (zooba) [inline Include/internal/pycore_atomic.h]: Are these deliberate changes? Or is it a merge that is accidentally undoing the fixes that @ericsnowcurrently made recently?

File: PC/layout/support/options.py [hunk 2]
```diff
@@ -55,6 +55,10 @@ def public(f):
         "help": "nuget package",
         "options": ["stable", "pip", "distutils", "dev", "props"],
     },
+    "iot": {
+        "help": "Windows IoT Core",
+        "options": ["stable", "pip"],
```
Comment (zooba) [inline PC/layout/support/options.py]: This seems unrelated to ctypes, yes?

File: PCbuild/python.props [hunk 3]
```diff
@@ -73,6 +73,7 @@
     <PyArchExt Condition="'$(ArchName)' == 'arm32'">-arm32</PyArchExt>
     
     <!-- Full path of the resulting python.exe binary -->
+    <PythonExe Condition="'$(PythonExe)' == '' and $(Platform) == 'ARM'">$(BuildPath32)python$(PyDebugExt).exe</PythonExe>
```
Comment (zooba) [inline PCbuild/python.props]: Can we use this condition when setting `BuildPath` rather than here?

**Review by zooba — COMMENTED** [review]
A few unrelated changes, it seems, but otherwise it looks okay. (Is there another PR I should be looking at first)

File: Include/internal/pycore_atomic.h (hunk 1 above)
Comment (paulmon) [reply Include/internal/pycore_atomic.h]: Yes this is deliberate.  It doesn't undo all of the @ericsnowcurrently changes.  Without this change the code evaluates to _InterlockedExchange_acq((volatile long*)&((&((&_PyRuntime.ceval.gil.last_holder)->_value))->_value) causing a build error for _M_ARM

File: PCbuild/python.props (hunk 3 above)
Comment (paulmon) [reply PCbuild/python.props]: No, I don't think so, but there is probably a better way to do what this is doing.   
Without this change the fragment
**"$(PythonExe)" "$(PySourcePath)PC\layout"** in  **Tools\nuget\make_pkg.proj** fails.  I would like to be able to produce an pythonarm32 nuget package to support cross-compiling python packages.

**Comment by paulmon** [comment]
> A few unrelated changes, it seems, but otherwise it looks okay. (Is there another PR I should be looking at first)

This PR adds libffi support for arm32 on windows.  https://github.com/python/cpython-source-deps/pull/9

File: PCbuild/python.props (hunk 3 above)
Comment (paulmon) [reply PCbuild/python.props]: This (or whatever replaces it) should be moved to a different PR for layout/nuget changes

File: PC/layout/support/options.py (hunk 2 above)
Comment (paulmon) [reply PC/layout/support/options.py]: Yes.  It is a part of my test requirements for ARM32, but it can be seperated from ctypes changes.  I'll make the layout changes a seperate PR.

File: Python/ceval.c [hunk 4]
```diff
@@ -1733,6 +1733,14 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
                 exc = POP(); /* exc */
                 /* fall through */
             case 0:
+#if defined(_M_ARM) && !defined(DEBUG)
```
Comment (paulmon) [inline Python/ceval.c]: I think this link says this is fixed in VS 15.8.  Does that seem right?
https://developercommunity.visualstudio.com/content/problem/345767/arm-optimization-change-function-parameter-data-al.html

File: Python/ceval.c [hunk 5]
```diff
@@ -1733,6 +1733,15 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
                 exc = POP(); /* exc */
                 /* fall through */
             case 0:
+#if defined(_M_ARM) && !defined(DEBUG) && (_MSC_VER < 1915)
```
Comment (paulmon) [inline Python/ceval.c]: I have confirmed that the arm optimization bug is fixed on newer builds of visual studio. 
Does it make sense to check for older versions of the tools with a #if like this or should this block of code just go away?

File: Include/internal/pycore_atomic.h (hunk 1 above)
Comment (paulmon) [reply Include/internal/pycore_atomic.h]: @zooba 
Also, see line 277.  This change makes arm/arm64 match what was done for x86/x64 in @ericsnowcurrently last change.

File: Include/pyport.h [hunk 6]
```diff
@@ -406,7 +406,7 @@ extern "C" {
 #endif
 
 /* get and set x87 control word for VisualStudio/x86 */
-#if defined(_MSC_VER) && defined(_M_IX86) /* x87 only supported in x86 */
+#if defined(_MSC_VER) && !defined(_WIN64) && !defined(_M_ARM) /* x87 not supported in 64-bit or ARM */
```
Comment (paulmon) [inline Include/pyport.h:409]: @zooba Do you think this change and the one in pythonrun.h are related enough to the ctypes changes?
Or should they be seperated into a different PR?

File: Include/internal/pycore_atomic.h (hunk 1 above)
Comment (paulmon) [reply Include/internal/pycore_atomic.h]: moved to a seperate PR

File: PC/layout/support/options.py (hunk 2 above)
Comment (paulmon) [reply PC/layout/support/options.py]: moved to a seperate PR

File: PCbuild/python.props (hunk 3 above)
Comment (paulmon) [reply PCbuild/python.props]: This was moved to the arm32 nuget PR and removed from this PR

File: Python/ceval.c (hunk 4 above)
Comment (zooba) [reply Python/ceval.c]: Either fixed in 15.9 or 16.0 - "Fixed pending release" means in a release that hasn't gone out yet, but it was set a while ago so maybe it just wasn't updated properly (they're mostly updated by an automatic process using internal bugs, so can be missed if the links aren't right)

File: Python/ceval.c (hunk 5 above)
Comment (zooba) [reply Python/ceval.c]: Let's just remove it.

File: Include/pyport.h (hunk 6 above)
Comment (zooba) [reply Include/pyport.h:409]: They're fine here I think

File: PC/layout/main.py [hunk 7]
```diff
@@ -66,18 +66,6 @@
 TOOLS_DIRS = FileNameSet("scripts", "i18n", "pynche", "demo", "parser")
 TOOLS_FILES = FileSuffixSet(".py", ".pyw", ".txt")
 
-def copy_if_modified(src, dest):
```
Comment (zooba) [inline PC/layout/main.py]: Did you mean to revert your changes to this file?

**Comment by zooba** [comment]
I'm running a libffi build with the ARM32 binaries now. Hopefully I'll get that done this evening.

**Comment by zooba** [comment]
@paulmon The ARM32 build failed - apparently the `-marm` option is unknown? See the logs starting at line 2456 of the "Install cygwin and build" step at https://dev.azure.com/python/cpython/_build/results?buildId=40818&_a=summary&view=logs

Any ideas?

File: PC/layout/main.py (hunk 7 above)
Comment (paulmon) [reply PC/layout/main.py]: No.  I meant to rebase on top of the layout changes in the other PR. I'll fix this

**Comment by paulmon** [comment]
> @paulmon The ARM32 build failed - apparently the `-marm` option is unknown? 
I think you still need to merge https://github.com/python/cpython-source-deps/pull/9 before building

File: Python/ceval.c (hunk 4 above)
Comment (paulmon) [reply Python/ceval.c]: Removed optimization workaround

File: PC/layout/main.py (hunk 7 above)
Comment (paulmon) [reply PC/layout/main.py]: fixed

**Comment by zooba** [comment]
@paulmon How did this pass without the libffi build? Which PR requires the build?

**Comment by paulmon** [comment]
Do you mean how did the CI build pass? The CI build passes because there is no arm32 in the strategy matrix in pr.yml (also ci.yml).  If arm was in the ci.yml.  

I think the CI/PR pipelines will need to skip running tests for ARM builds.  

Also, if you add ARM to the CI build right now the build will break on this: https://github.com/python/cpython/pull/12665

**Comment by paulmon** [comment]
One more thing.  To build ARM from master --no-tkinter and --no-ssl are both needed as well.  I can propose changes to the CI.yml files if you think that would be helpful.

**Comment by zooba** [comment]
> Do you mean how did the CI build pass? The CI build passes because there is no arm32 in the strategy matrix in pr.yml (also ci.yml).

Ah yeah, that makes sense :)

Let's just get the buildbot plugged in. We're clearly not going to run tests from any of our CI providers just yet.
